naga/src/proc/constant_evaluator.rs

```diff
@@ -2103,6 +2106,41 @@ impl<'a> ConstantEvaluator<'a> {
         Ok(Expression::Compose { ty, components })
     }
 
+    fn relational(
+        &mut self,
+        fun: RelationalFunction,
+        arg: Handle<Expression>,
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        let arg = self.eval_zero_value_and_splat(arg, span)?;
+        match fun {
+            RelationalFunction::All | RelationalFunction::Any => match self.expressions[arg] {
+                Expression::Literal(Literal::Bool(_)) => Ok(arg),
+                Expression::Compose { ty, ref components }
+                    if matches!(self.types[ty].inner, TypeInner::Vector { .. }) =>
+                {
+                    let components =
+                        crate::proc::flatten_compose(ty, components, self.expressions, self.types)
+                            .map(|component| match self.expressions[component] {
+                                Expression::Literal(Literal::Bool(val)) => Ok(val),
+                                _ => Err(ConstantEvaluatorError::InvalidRelationalArg(fun)),
+                            })
+                            .collect::<Result<ArrayVec<bool, { crate::VectorSize::MAX }>, _>>()?;
+                    let result = match fun {
+                        RelationalFunction::All => components.iter().all(|c| *c),
+                        RelationalFunction::Any => components.iter().any(|c| *c),
+                        _ => unreachable!(),
+                    };
+                    self.register_evaluated_expr(Expression::Literal(Literal::Bool(result)), span)
+                }
+                _ => Err(ConstantEvaluatorError::InvalidRelationalArg(fun)),
+            },
+            _ => Err(ConstantEvaluatorError::NotImplemented(format!(
+                "{fun:?} built-in function"
+            ))),
+        }
+    }
+
```

Comment:
question: Are the component-wise macros not helpful for making a more expressive implementation here?

Reply:
It looked to me like they were useful when applying an operation to each component, but I didn't see how they would help me here where we need to iterate over all components. I may be missing something though

Reply:
I tried to tackle this once but it got staled: https://github.com/gfx-rs/wgpu/pull/6213

Reply:
@sagudev: Mmm, I'd forgotten about that PR, sorry! We don't really re-triage PRs that stay in draft during our maintainer sync meetings. 😅

This PR is scoped to only `any` and `all`, which also makes this significantly easier to review.